landlab/components/groundwater/dupuit_percolator.py

```diff
@@ -361,6 +368,20 @@ def __init__(
         self.courant_coefficient = courant_coefficient
         self.vn_coefficient = vn_coefficient
 
+        if k_func is not None:
+            if (
+                not isinstance(k_func(self._grid), np.ndarray)
+                and len(k_func(self._grid)) == self._grid.number_of_links
+            ):
+                raise TypeError(
+                    """The k_func should be a function that takes a ModelGrid
+                    and returns an array of length number_of_links."""
+                )
```

Comment:
Does this mean that the run_one_step methods still need a line `self.K = self.K` ?

Reply:
I think my one hesitation here would be that it it would be checking whether it is the right type, etc. every time it is called to update, rather than just in init. Does that make sense?

Reply:
Good point. I think the following would address this issue.

```python
    @K.setter
    def K(self, new_val):
        """set hydraulic conductivity at link (m/s)"""
        if callable(new_val):
             self._kfunc=True:
             # do tests on new_val here. 
             self._func = new_val
             self._K=return_array_at_link(self._grid, self._func(self._grid))
       else:
            self._kfunc=False
            self._K = return_array_at_link(self._grid, new_val)

    @property
    def K(self):
        """hydraulic conductivity at link (m/s)"""
        if self._kfunc:
            self._K=return_array_at_link(self._grid, self._func(self._grid))
    return self._K        

```

Reply:
Thanks! Addressed in the most recent commits.